netebpfext/net_ebpf_ext_sock_ops.c

```diff
@@ -203,42 +203,35 @@ _net_ebpf_extension_sock_ops_on_client_detach(_In_ const net_ebpf_extension_hook
     net_ebpf_extension_sock_ops_wfp_filter_context_t* filter_context =
         (net_ebpf_extension_sock_ops_wfp_filter_context_t*)net_ebpf_extension_hook_client_get_provider_data(
             detaching_client);
-    net_ebpf_extension_flow_context_parameters_t* flow_parameters_array = NULL;
-    uint32_t flow_count = 0;
-    uint32_t flow_index = 0;
     KIRQL irql;
-    bool lock_held = FALSE;
+    LIST_ENTRY local_list_head;
+
     ASSERT(filter_context != NULL);
+    InitializeListHead(&local_list_head);
     net_ebpf_extension_delete_wfp_filters(filter_context->base.filter_ids_count, filter_context->base.filter_ids);
 
     KeAcquireSpinLock(&filter_context->lock, &irql);
-    lock_held = TRUE;
-    flow_count = filter_context->flow_context_list.count;
-    if (flow_count > 0) {
-        flow_parameters_array = (net_ebpf_extension_flow_context_parameters_t*)ExAllocatePoolUninitialized(
-            NonPagedPoolNx,
-            sizeof(net_ebpf_extension_flow_context_parameters_t) * flow_count,
-            NET_EBPF_EXTENSION_POOL_TAG);
-        if (flow_parameters_array == NULL)
-            goto Exit;
-        // Remove any remaining flow contexts and copy their flow parameters.
-        while (!IsListEmpty(&filter_context->flow_context_list.list_head)) {
-            LIST_ENTRY* entry = RemoveHeadList(&filter_context->flow_context_list.list_head);
-            InitializeListHead(entry);
-            net_ebpf_extension_sock_ops_wfp_flow_context_t* flow_context =
-                CONTAINING_RECORD(entry, net_ebpf_extension_sock_ops_wfp_flow_context_t, link);
-            flow_context->client_detached = TRUE;
-            _Analysis_assume_(flow_index < flow_count);
-            flow_parameters_array[flow_index++] = flow_context->parameters;
-        }
+    if (filter_context->flow_context_list.count > 0) {
```

Comment:
was the intent to not have allocation within the lock? Where was the leak here?

Reply:
This function was allocating an array in the client detach path. If the allocation failed, this function failed to call `FwpsFlowRemoveContext` for all the flows and returned an error. That caused memory leak.

While fixing, I realized we dont need to actually allocate memory and can also snap the list under lock and later iterate over the list.